include/execpools/thread_pool_base.hpp

```diff
@@ -0,0 +1,485 @@
+              receiver_of<Receiver, completion_signatures<Self, stdexec::env_of_t<Receiver>>>
+            static auto connect(Self&& self, Receiver rcvr) //
+            noexcept(stdexec::__nothrow_constructible_from<
+                     bulk_op_state_t<Self, Receiver>,
+                     DerivedPoolType&,
+                     Shape,
+                     Fun,
+                     Sender,
+                     Receiver>) -> bulk_op_state_t<Self, Receiver> {
+            return bulk_op_state_t<Self, Receiver>{
+              self.pool_,
+              self.shape_,
+              self.fun_,
+              static_cast<Self&&>(self).sndr_,
+              static_cast<Receiver&&>(rcvr)};
+          }
+
+          template <stdexec::__decays_to<__t> Self, class... Env>
+          static auto
+            get_completion_signatures(Self&&, Env&&...) -> completion_signatures<Self, Env...> {
+            return {};
+          }
+
+          template <stdexec::__forwarding_query Tag, class... As>
+            requires stdexec::__callable<Tag, const Sender&, As...>
+          auto query(Tag, As&&... as) const
+            noexcept(stdexec::__nothrow_callable<Tag, const Sender&, As...>) -> decltype(auto) {
+            return Tag()(sndr_, static_cast<As&&>(as)...);
+          }
+
+          auto get_env() const noexcept -> const __t& {
+            return *this;
+          }
+        };
+      };
+
+      template <stdexec::sender Sender, std::integral Shape, class Fun>
+      using bulk_sender_t =
+        stdexec::__t<bulk_sender<stdexec::__id<stdexec::__decay_t<Sender>>, Shape, Fun>>;
+
+      STDEXEC_MEMFN_FRIEND(bulk);
+
+      template <stdexec::sender S, std::integral Shape, class Fun>
+        STDEXEC_MEMFN_DECL(auto bulk)(this const scheduler& sch, S&& sndr, Shape shape, Fun fun) noexcept
+        -> bulk_sender_t<S, Shape, Fun> {
+        return bulk_sender_t<S, Shape, Fun>{
+          *sch.pool_, static_cast<S&&>(sndr), shape, static_cast<Fun&&>(fun)};
+      }
+
+      [[nodiscard]]
+      constexpr auto
+        forward_progress_guarantee() const noexcept -> stdexec::forward_progress_guarantee {
+        return pool_->forward_progress_guarantee();
+      }
+
+      friend thread_pool_base;
+
+      explicit scheduler(DerivedPoolType& pool) noexcept
+        : pool_(&pool) {
+      }
+
+      DerivedPoolType* pool_;
+
+     public:
+      using __t = scheduler;
+      using __id = scheduler;
+      auto operator==(const scheduler&) const -> bool = default;
+
+      constexpr auto query(stdexec::get_forward_progress_guarantee_t) const noexcept
+        -> stdexec::forward_progress_guarantee {
+        return forward_progress_guarantee();
+      }
+
+      auto schedule() const noexcept -> sender {
+        return sender{*pool_};
+      }
+    };
+
+    [[nodiscard]]
+    auto get_scheduler() noexcept -> scheduler {
+      return scheduler{static_cast<DerivedPoolType&>(*this)};
+    }
+
+    /* Is this even needed? Looks like no.
```

Comment:
I left this comment in here I think... I didn't get an answer to the question and forgot to remove it.